meteor/sofie-server/model/entities/rundown.ts

```diff
@@ -159,11 +211,113 @@ export class Rundown extends BasicRundown {
+	}
+
+	private findOldInfinitePiecesNotOnLayers(layers: Set<string>): Piece[] {
+		return Array.from(this.infinitePieces.values()).filter((oldPiece) => {
+			return !layers.has(oldPiece.layer)
+		})
+	}
+
+	private isPieceOutlived(piece: Piece): boolean {
+		switch (piece.pieceLifespan) {
+			case PieceLifespan.WITHIN_PART: {
+				// Not an infinite, so we don't care about it and just mark it as outlived.
+				return true
+			}
+			case PieceLifespan.STICKY_UNTIL_RUNDOWN_CHANGE: {
+				// Since we are in the context of a Rundown then the Piece will never be able to leave the Rundown, so the Piece is NOT outlived.
+				return false
+			}
+			case PieceLifespan.STICKY_UNTIL_SEGMENT_CHANGE: {
+				// If the Piece belongs to the active Segment, then the Piece is NOT outlived.
+				return !this.activeSegment.doesPieceBelongToSegment(piece)
+			}
+			case PieceLifespan.SPANNING_UNTIL_RUNDOWN_END:
+			case PieceLifespan.SPANNING_UNTIL_SEGMENT_END: {
+				// We always mark SPANNING as outlived because even it if isn't we need to check if there is another SPANNING Piece between this Piece and the active Part.
+				return true
+			}
+			default: {
+				throw new UnsupportedOperation(
+					`{${piece.pieceLifespan}} is not supported. Are you missing an implementation?`
+				)
+			}
+		}
+	}
+
+	private resetOutlivedInfinitePieces(piecesThatHasNotBeenOutlived: Piece[]): void {
+		const pieceIdsThatHasNotBeenOutlived: string[] = piecesThatHasNotBeenOutlived.map((piece) => piece.id)
+		Array.from(this.infinitePieces.values())
+			.filter((piece) => !pieceIdsThatHasNotBeenOutlived.includes(piece.id))
+			.forEach((piece) => piece.resetExecutedAt())
+	}
+
+	private addPiecesToLayers(pieces: Piece[], layersWithPieces: Map<string, Piece>): Map<string, Piece> {
+		pieces.forEach((piece) => {
+			if (layersWithPieces.has(piece.layer)) {
+				throw new Error(
+					`${piece.pieceLifespan}: Trying to add an infinite Piece to a layer that already have an infinite Piece`
+				)
+			}
+			layersWithPieces.set(piece.layer, piece)
+		})
+		return layersWithPieces
+	}
+
+	private addSpanningPiecesNotOnLayersFromActiveSegment(layersWithPieces: Map<string, Piece>): Map<string, Piece> {
+		const piecesToAdd: Piece[] = this.activeSegment.getFirstSpanningPieceForEachLayerBeforePart(
+			this.activePart,
+			new Set(layersWithPieces.keys())
+		)
+		return this.addPiecesToLayers(piecesToAdd, layersWithPieces)
+	}
+
+	private addSpanningPiecesNotOnLayersFromPreviousSegments(layersWithPieces: Map<string, Piece>): Map<string, Piece> {
+		const indexOfActiveSegment = this.segments.findIndex((segment) => segment.id === this.activeSegment.id)
+		for (let i = indexOfActiveSegment - 1; i >= 0; i--) {
+			const piecesSpanningSegment: Piece[] = this.segments[i].getFirstSpanningRundownPieceForEachLayerForAllParts(
+				new Set(layersWithPieces.keys())
+			)
+			layersWithPieces = this.addPiecesToLayers(piecesSpanningSegment, layersWithPieces)
+		}
+		return layersWithPieces
+	}
```

Comment:
This would be nice to refactor to a less imperative (and shadowing) style.

Reply:
Suggestions are most welcome :)
We should have a chat about it